### PR TITLE
fix unlisted upload validation

### LINKED ISSUE #9448: Unlisted extension submission is stuck in validation mode
STR

1. Log in to AMO -dev
2. Submit a new unlisted extension
3. Check the validation results

Actual result:
Submission is stuck in validation mode

Expected result:
Add-on is passing validation

Notes;
- reproduced on AMO -dev with FF62, Win10x64
- In my previous submissions, I was able to reproduce with listed versions as well, but in my last attempts I was able to constantly reproduce only with unlisted versions

image

### PR DESCRIPTION
**Pull request by eviljeff**
fixes #9448

### CODE REVIEW

File: src/olympia/devhub/views.py [hunk 1]
```diff
@@ -580,10 +580,10 @@ def handle_upload(filedata, request, channel, app_id=None, version_id=None,
                 user_agent=request.META.get('HTTP_USER_AGENT'),
                 referrer=request.META.get('HTTP_REFERER'),
                 upload=upload)
-            akismet_checks = comment_check.si(
-                [report.id for report in akismet_reports])
         else:
-            akismet_checks = None
+            akismet_reports = []
+        akismet_checks = comment_check.si(
```
Comment (EnTeQuAk) [inline src/olympia/devhub/views.py]: you'll probably don't want to do that if `akismet_reports` is empty I guess?

**Comment by EnTeQuAk** [comment]
As mentioned in IRC: We'll want to force the `comment_check` task to be executed in the `devhub` queue when running in the validation chain (http://docs.celeryproject.org/en/latest/userguide/calling.html#routing-options) and `comment_check` also should not ignore task results (which we do by default) http://docs.celeryproject.org/en/latest/userguide/canvas.html#important-notes

File: src/olympia/devhub/views.py (hunk 1 above)
Comment (eviljeff) [reply src/olympia/devhub/views.py]: That's the part you _have_ to do to fix this.  No task means the wrong number of parameters are passed on to `validate_file_path` or `validate_file`.  `chain()` doesn't do this.  (`group()`, which makes a chord, does in eager mode but seemingly not in non-eager)

**Comment by eviljeff** [comment]
> comment_check also should not ignore task results (which we do by default) 

what actually needs changing for this?

File: src/olympia/devhub/views.py (hunk 1 above)
Comment (EnTeQuAk) [reply src/olympia/devhub/views.py]: Let's add a very verbose and explaining comment why this is necessary, it broke before and I fear if we touch it again in a few months time we are doomed to do that again. Both here, and in `devhub.tasks` given that this is generic functionality that's potentially exposed to every `validation()` caller.

**Comment by EnTeQuAk** [comment]
```diff
diff --git a/src/olympia/lib/akismet/tasks.py b/src/olympia/lib/akismet/tasks.py
index a106429858..434abfab99 100644
--- a/src/olympia/lib/akismet/tasks.py
+++ b/src/olympia/lib/akismet/tasks.py
@@ -10,7 +10,7 @@ def submit_to_akismet(report_ids, submit_spam):
         report.submit_spam() if submit_spam else report.submit_ham()


-@task
+@task(ignore_result=False)
 def comment_check(report_ids):
     reports = AkismetReport.objects.filter(id__in=report_ids)
     return [report.comment_check() for report in reports]
```

File: src/olympia/devhub/tasks.py [hunk 2]
```diff
@@ -72,8 +72,21 @@ def validate(file_, listed=None, subtask=None, synchronous=False,
...
+        # for pretask but if this is changed it WILL break on dev/stage/prod.
+        #
+        # We HAVE to have a task at the start of the chain that returns a
+        # result (ignore_result=False) as the result from pretask is passed
+        # down to the next task in the chain, validation.task, which is wrapped
+        # in validation_task.  If there no task then validation_task fails with
+        # mismatched arguments (same with something like an empty celery.group)
+        #
+        # In a ideal world this would all be rewritten with chords and
+        # parallel tasks in groups and magic.
```
Comment (EnTeQuAk) [inline src/olympia/devhub/tasks.py]: Thanks! :heart:

**Pull request merged**